Remove interconnected IMA, Thrift, and ClientImpl non-public methods

SethSmucker opened the pull request:
InMemoryAccumulo, Thrift, and ClientImpl are non-public methods in Accumulo 2.1. All three needed to be removed in a group since they were reliant on each other. They have been replaced with public variations where applicable, or new methods have been created to handle the job they used to do.

Part of #2443

dlmarion commented on warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/job/BulkIngestMapFileLoader.java:
```diff
@@ -703,42 +703,22 @@ private Path distCpDirectory(Path jobDirectory) throws Exception {
-        ClientContext context = (ClientContext) accumuloClient;
-        try {
-            client = ThriftClientTypes.MANAGER.getConnection(context);
-            ManagerMonitorInfo mmi = client.getManagerStats(null, context.rpcCreds());
-            Map<String,TableInfo> tableStats = mmi.getTableMap();
+        log.log(level, "There are " + majminC + " compactions currently running or queued.");
 
-            for (java.util.Map.Entry<String,TableInfo> e : tableStats.entrySet()) {
-                majC += e.getValue().getMajors().getQueued();
-                majC += e.getValue().getMajors().getRunning();
```
This should still work with Accumulo 2.1. Are these changes targeting Accumulo 4.0?

foster33 replied:
Yes, 4.0

SethSmucker committed:
Remove interconnected IMA, Thrift, and ClientImpl non-public methods

- Remove extends ClientContext from InMemoryAccumuloClient
- Remove extends Connector from InMemoryConnector
- Remove implements Instance from InMemoryInstance
- Delete InMemoryClientInfo.java (used non-public Credentials/ClientInfoImpl)
- Clean up BulkInputFormat TabletLocator usage
- Update PushdownScheduler to remove ClientContext usage

Part of #2443

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/ima-remove-the-rest

SethSmucker commented:
## Stacked PR Notice

This PR is the **BASE** of a stacked PR chain. The following PRs depend on this one and share files:

| PR | Title | Status |
|----|-------|--------|
| **#3253** | Remove interconnected IMA, Thrift, and ClientImpl | **THIS PR (BASE)** |
| #3345 | Replace InMemoryInstance with InMemoryAccumulo | Stacked on this |
| #3227 | Remove non-public clientImpl.* usages | DRAFT, stacked on this |

**Shared files:**
- `InMemoryAccumuloClient.java`
- `InMemoryClientInfo.java`
- `InMemoryConnector.java`
- `InMemoryInstance.java`
- `BulkInputFormat.java`

**IMPORTANT:** This PR must merge FIRST. After merging, #3345 and #3227 will need to be updated with integration.

SethSmucker committed:
Fix code formatting issues

Apply formatter, sortpom, and impsort fixes to pass CI check.

SethSmucker committed:
Remove unused getClientContext utility from AccumuloConnectionFactory

The getClientContext method used non-public ClientContext API and was
only called from PushdownScheduler. Since PushdownScheduler now uses
the public tableOperations().tableIdMap() API instead, this utility
method is no longer needed.

Supersedes PR #3340.

SethSmucker committed:
Clarify intentionally unsupported methods in InMemoryAccumulo

SethSmucker committed:
Merge with integration

ivakegg committed:
Add imports back in

ivakegg committed:
Added in too many

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/ima-remove-the-rest

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/ima-remove-the-rest

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/ima-remove-the-rest

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/ima-remove-the-rest

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/ima-remove-the-rest

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/ima-remove-the-rest

SethSmucker committed:
Merge branch 'integration' into task/ima-remove-the-rest

ivakegg commented on warehouse/core/src/main/java/datawave/mr/bulk/BulkInputFormat.java:
```diff
@@ -1105,10 +1104,10 @@ public List<InputSplit> getSplits(JobContext job) throws IOException {
                             if (tableId == null)
-                                tableId = context.getTableId(tableName);
-                            if (!context.tableNodeExists(tableId))
+                                tableId = TableId.of(client.tableOperations().tableIdMap().get(tableName));// TABLE ID todo:ensure this is correct
```
Are these TODO's still valid?

apmoriarty replied:
Second this

lbschanno replied:
Third this.

ddanielr commented on warehouse/core/src/main/java/datawave/mr/bulk/BulkInputFormat.java:
```diff
@@ -1105,10 +1104,10 @@ public List<InputSplit> getSplits(JobContext job) throws IOException {
                             if (tableId == null)
-                                tableId = context.getTableId(tableName);
-                            if (!context.tableNodeExists(tableId))
+                                tableId = TableId.of(client.tableOperations().tableIdMap().get(tableName));// TABLE ID todo:ensure this is correct
+                            if (!client.tableOperations().tableIdMap().containsKey(tableName))// CHECK IF TABLE ID EXISTS //todo same
                                 throw new TableDeletedException(tableId.canonical());
-                            if (context.getTableState(tableId) == TableState.OFFLINE)
+                            if (!client.tableOperations().isOnline(tableName)) // CHECK IF TABLE ID OFFLInE //todo same
                                 throw new TableOfflineException("Table (" + tableId.canonical() + ") is offline");
```
The `tableId` seems to only be used for these exception messages. So that variable could be removed and this whole code block could be simplified to the following:
```
 if (!client.tableOperations().isOnline(tableName)) {
     throw new TableOfflineException((TableId) null, tableName);
}
```

The `isOnline` method checks if the table exists, finds the tableId, and then checks if the table is online via the tableID.

It also runs namespace validation which the `.tableIdMap().get()` method does not do. This is more important if you start to leverage accumulo namespaces (e.g `datawave.metadata`)


if the tableId is null in the TableOfflineException then the exception message reports an unknown id
`Table test (<unknown id>) is offline` 

The only behavior change is that if `isOnline` can't find a table id for the `tableName` then it will throw a `TableNotFound` exception instead of a `TableDeletedException`.

However all exceptions are caught and re-thrown as an `IOException` so I don't think the exception type changing will matter.

lbschanno committed:
Merge branch 'integration' into task/ima-remove-the-rest

lbschanno commented on core/in-memory-accumulo/src/main/java/datawave/accumulo/inmemory/InMemoryConnector.java:
```diff
@@ -25,130 +25,99 @@
     }
 
-    @Override
     public ConditionalWriter createConditionalWriter(String tableName, ConditionalWriterConfig config) throws TableNotFoundException {
```
I would move this documentation to the actual JavaDoc for the method.

lbschanno commented on core/in-memory-accumulo/src/main/java/datawave/accumulo/inmemory/InMemoryAccumuloClient.java:
```diff
@@ -36,36 +36,47 @@
+    }
+
+    public ConditionalWriter createConditionalWriter(String tableName) throws TableNotFoundException {
+        return this.createConditionalWriter(tableName, (ConditionalWriterConfig) null);
```
No need to cast a null value to ConditionalWriterConfig here.

lbschanno commented on core/in-memory-accumulo/src/main/java/datawave/accumulo/inmemory/InMemoryAccumuloClient.java:
```diff
@@ -36,36 +36,47 @@
+        return this.createConditionalWriter(tableName, (ConditionalWriterConfig) null);
+    }
+
+    private void ensureOpen() {
```
This method is showing as unused. Is it still necessary?

lbschanno commented on core/in-memory-accumulo/src/main/java/datawave/accumulo/inmemory/InMemoryAccumuloClient.java:
```diff
@@ -36,36 +36,47 @@
     String username;
     private final InMemoryAccumulo acu;
 
+    private ConditionalWriterConfig conditionalWriterConfig;
```
This ConditionalWriterConfig is showing as unused. Is it still necessary?

lbschanno commented on core/in-memory-accumulo/src/main/java/datawave/accumulo/inmemory/InMemoryAccumuloClient.java:
```diff
@@ -36,36 +36,47 @@
     private final InMemoryAccumulo acu;
 
+    private ConditionalWriterConfig conditionalWriterConfig;
+    private volatile boolean closed = false;
```
This can be final.

lbschanno commented on core/in-memory-accumulo/src/main/java/datawave/accumulo/inmemory/InMemoryConnector.java:
```diff
@@ -25,130 +25,99 @@
     private final InMemoryAccumulo acu;
-    private final Instance instance;
 
     InMemoryConnector(String username, InMemoryInstance instance) throws AccumuloSecurityException {
```
AccumuloSecurityException here is not actually thrown.

lbschanno commented on core/in-memory-accumulo/src/main/java/datawave/accumulo/inmemory/InMemoryAccumuloClient.java:
```diff
@@ -36,36 +36,47 @@
+    private ConditionalWriterConfig conditionalWriterConfig;
+    private volatile boolean closed = false;
+
     public InMemoryAccumuloClient(String username, InMemoryInstance instance) throws AccumuloSecurityException {
```
AccumuloSecurityException is not actually being thrown.

lbschanno commented on core/in-memory-accumulo/src/main/java/datawave/accumulo/inmemory/InMemoryAccumuloClient.java:
```diff
@@ -36,36 +36,47 @@
 
+    @Override
+    public ConditionalWriter createConditionalWriter(String tableName, ConditionalWriterConfig config) throws TableNotFoundException {
+        // Intentionally unsupported - required for AccumuloClient interface but not used by DataWave
```
Would like to see this moved to the JavaDoc.

lbschanno commented on core/in-memory-accumulo/src/main/java/datawave/accumulo/inmemory/InMemoryAccumuloClient.java:
```diff
@@ -36,36 +36,47 @@
     }
 
+    @Override
+    public ConditionalWriter createConditionalWriter(String tableName, ConditionalWriterConfig config) throws TableNotFoundException {
```
TableNotFoundException is not thrown here.

lbschanno commented on core/in-memory-accumulo/src/main/java/datawave/accumulo/inmemory/InMemoryAccumuloClient.java:
```diff
@@ -156,15 +167,9 @@ public NamespaceOperations namespaceOperations() {
     @Override
     public ReplicationOperations replicationOperations() {
-        // TODO add implementation
+        // Intentionally unsupported - required for AccumuloClient interface but not used by DataWave
```
Would like this moved to the JavaDoc.

SethSmucker committed:
Address lbschanno and ddanielr review feedback

InMemoryAccumuloClient:
- Remove unused ConditionalWriterConfig field and ensureOpen method
- Remove AccumuloSecurityException from constructor (not thrown)
- Remove unused createConditionalWriter(String) overload
- Remove TableNotFoundException from createConditionalWriter (not thrown)
- Move inline comments to javadoc

InMemoryConnector:
- Remove AccumuloSecurityException from constructor (not thrown)
- Move inline comment to javadoc

BulkInputFormat:
- Simplify online check per ddanielr: replace tableId/tableIdMap/isOnline
  block with single isOnline(tableName) call
- Remove unused tableId variable and TableDeletedException import

Part of #2443

SethSmucker committed:
Remove unused AccumuloSecurityException import from InMemoryConnector

SethSmucker committed:
Restore createConditionalWriter(String) required by AccumuloClient interface

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/ima-remove-the-rest

lbschanno committed:
Merge branch 'integration' into task/ima-remove-the-rest

foster33 committed:
Merge branch 'integration' into task/ima-remove-the-rest

foster33 reviewed (commented):
When using the changes to `in-memory-accumulo` introduced in this PR, there is a compilation failure:
```
warehouse/query-core/src/test/java/datawave/query/util/MockMetadataHelper.java:[89,11] exception org.apache.accumulo.core.client.AccumuloSecurityException is never thrown in body of corresponding try statement
```
Unsure if this is related to this PR or if it is an issue with the current SNAPSHOT version

SethSmucker committed:
Merge remote-tracking branch 'origin/task/ima-remove-the-rest' into task/ima-remove-the-rest

SethSmucker committed:
Make username field final, remove unnecessary throws declarations

SethSmucker committed:
Bump in-memory-accumulo to 4.0.6-SNAPSHOT and fix MockMetadataHelper

Constructor no longer declares AccumuloSecurityException, so remove
the dead try/catch in MockMetadataHelper.getClient().

SethSmucker committed:
Merge branch 'integration' into task/ima-remove-the-rest

SethSmucker commented:
> When using the changes to `in-memory-accumulo` introduced in this PR, there is a compilation failure:
> 
> ```
> warehouse/query-core/src/test/java/datawave/query/util/MockMetadataHelper.java:[89,11] exception org.apache.accumulo.core.client.AccumuloSecurityException is never thrown in body of corresponding try statement
> ```
> 
> Unsure if this is related to this PR or if it is an issue with the current SNAPSHOT version

4.0.4 is still being used on integration, so until we update to 4.0.6-SNAPSHOT we won't be able to use any of the changes here even though we build them.

> When using the changes to `in-memory-accumulo` introduced in this PR, there is a compilation failure:
> 
> ```
> warehouse/query-core/src/test/java/datawave/query/util/MockMetadataHelper.java:[89,11] exception org.apache.accumulo.core.client.AccumuloSecurityException is never thrown in body of corresponding try statement
> ```
> 
> Unsure if this is related to this PR or if it is an issue with the current SNAPSHOT version

I just swapped over from 4.0.4 to 4.0.6-SNAPSHOT, it should be good to go now!

SethSmucker committed:
Add /etc/hadoop/conf symlink in quickstart Docker image

Datawave's server-side query iterators load Hadoop configuration from
/etc/hadoop/conf when they run inside the Accumulo TabletServer. That
path does not exist in the quickstart container — Hadoop is installed
under /opt/datawave/contrib/datawave-quickstart/hadoop and HADOOP_CONF_DIR
points there, not at /etc/hadoop/conf. Without the path, every
event-table scan throws FileNotFoundException server-side and the
query service silently returns 0 events.

Symlink /etc/hadoop/conf to the installed Hadoop config directory so the
expected path resolves. Verified locally: docker/scripts/testAll.sh goes
from 5/13 to 13/13 passing.

SethSmucker committed:
Add quickstart dependency for services that connect to ZooKeeper or Hadoop

Nine services in docker-compose.yml use ZOOKEEPER_HOST or HADOOP_HOST
but did not declare a depends_on for the quickstart container, which
hosts ZooKeeper, HDFS, and Accumulo. Only metrics correctly declared
the dependency.

When quickstart's startup happens to take longer than the 60s ZooKeeper
connect timeout used by Spring services like audit's AccumuloAuditor
bean, those services exit 1 during init with "Failed to connect to
zookeeper (quickstart:2181) within 2x zookeeper timeout period 30000",
and docker compose aborts with "dependency failed to start: container
docker-audit-1 exited (1)". The race is environment-dependent: on
slower runners it is frequently lost, producing intermittent CI failures.

Declare depends_on quickstart (condition: service_healthy) for
annotation, accumulo, audit, dictionary, modification, query,
mapreduce-query, executor-pool1, and executor-pool2 so the startup is
correctly serialized behind the container that provides their backing
services.

SethSmucker committed:
Merge branch 'integration' into task/compose-tests-stability

SethSmucker committed:
Bump spring-boot-starter-datawave-query consumers to 1.0.11-SNAPSHOT

PR #3560 removed deprecated eventPerDayThreshold and shardsPerDayThreshold
placeholders from the starter source and from docker/config/application-query.yml,
but the released spring-boot-starter-datawave-query-1.0.10.jar still embeds
the old QueryLogicFactory.xml that references those placeholders. With no
fresh 1.0.11 release cut yet, every consumer of the 1.0.10 jar (query,
executor, mapreduce-query, cached-results) fails Spring init at startup
with:

  Caused by: java.lang.IllegalArgumentException: Could not resolve
  placeholder 'datawave.query.logic.logics.BaseEventQuery.eventPerDayThreshold'

This blocks the compose-tests workflow on every recent PR opened against
integration (#3559, #3562, #3563, #3565 all show the same FAILURE).

Bump <version.datawave.starter-query> from 1.0.10 to 1.0.11-SNAPSHOT in the
six consuming poms so the multi-module reactor builds the starter from
current source (which has #3560's removal) and the query/executor docker
images embed the fresh jar.

This is a stop-gap. The proper fix is running microservice-cascade-release.yml
starting from microservices/starters/query to publish a real 1.0.11; once
that lands, these SNAPSHOT pins should be replaced with the released version.

SethSmucker committed:
Merge task/compose-tests-stability into task/ima-remove-the-rest to validate PR independent of docker-compose CI bugs

foster33 committed:
Merge branch 'integration' into task/ima-remove-the-rest

foster33 committed:
Catch AccumuloSecurityException in MockMetadataHelper

foster33 committed:
Remove latest exception changes

foster33 committed:
Merge branch 'integration' into task/ima-remove-the-rest

foster33 committed:
Merge branch 'integration' into task/ima-remove-the-rest
